module/localisation/BallLocalisation/src/BallModel.hpp

```diff
@@ -38,37 +44,54 @@ namespace localisation {
 
         static constexpr size_t size = 2;
 
+        using StateVec = Eigen::Matrix<Scalar, size, 1>;
+        using StateMat = Eigen::Matrix<Scalar, size, size>;
+
         struct MeasurementType {
             struct BALL {};
         };
 
-        arma::vec2 processNoiseDiagonal;
-
+        // Local variables for this model. Set their values from config file
+        // Number of reset particles
+        int NUM_ROGUES;
+        // Range of reset particles
+        Eigen::Matrix<Scalar, 2, 1> resetRange;
+        // Diagonal noise matrix
+        Eigen::Matrix<Scalar, 2, 1> processNoiseDiagonal;
 
-        BallModel() : processNoiseDiagonal(arma::fill::eye) {}  // empty constructor
+        BallModel() : NUM_ROGUES(), resetRange(), processNoiseDiagonal() {}
 
-        arma::vec::fixed<size> timeUpdate(const arma::vec::fixed<size>& state, double deltaT);
-
-        arma::vec3 predictedObservation(const arma::vec::fixed<size>& state,
-                                        const message::support::FieldDescription& field,
-                                        const utility::math::matrix::Transform3D& Hcw) const;
+        StateVec time(const StateVec& state, const Scalar& /*deltaT*/) const {
+            return state;
+        }
 
-        arma::vec observationDifference(const arma::vec& measurement, const arma::vec3& rBCc) const;
+        Eigen::Matrix<Scalar, 3, 1> predict(const StateVec& state,
+                                            const message::support::FieldDescription& field,
+                                            const Eigen::Matrix<Scalar, 4, 4>& Hcw) const {
 
-        arma::vec::fixed<size> limitState(const arma::vec::fixed<size>& state) const;
+            const Eigen::Matrix<Scalar, 4, 1> rBWw(state[PX], state[PY], field.ball_radius, 1.0);
+            const Eigen::Matrix<Scalar, 3, 1> rBCc_cart((Hcw * rBWw).head(3));
```

Comment:
All of the ones where I have `.head(3)` instead of `.head<3>()` are because they don't compile with that syntax, without an intermediate typecast. In this instance, adding this suggestion results in this error (which is the error which happens with all of them):
```cpp
error: expected primary-expression before ‘)’ token
   73 |             const Eigen::Matrix<Scalar, 3, 1> rBCc_cart((Hcw * rBWw).head<3>());
```
If I change the line by wrapping the intermediate type, it compiles, but this seems less readable to me than what I have got
```cpp
const Eigen::Matrix<Scalar, 3, 1> rBCc_cart(Eigen::Vector4d(Hcw * rBWw).head<3>());
```
I am interested if you agree

Reply:
`.matrix().head<3>()`?

Reply:
Same error as without the wrap:
```cpp
 error: invalid operands of types ‘<unresolved overloaded function type>’ and ‘int’ to binary ‘operator<’
   73 |             const Eigen::Matrix<Scalar, 3, 1> rBCc_cart((Hcw * rBWw).matrix().head<3>());
```

Reply:
Thats a different error. Anyways, what about this
```
const​ Eigen::Matrix<Scalar, ​3​, ​1​> ​rBCc_cart​ = (Hcw * rBWw).head<3>();
```

Reply:
tbf, it was the same as another error I had without a wrap when I tried something else ;p

This suggestion has the same error as the one I just commented with :ppp
```cpp
error: invalid operands of types ‘<unresolved overloaded function type>’ and ‘int’ to binary ‘operator<’
   73 |             const Eigen::Matrix<Scalar, 3, 1> rBCc_cart = (Hcw * rBWw).head<3>();
```

Reply:
I am thinking dependant scope is being an ass to us. Try `.template head<3>();`

Reply:
It builds, but I am not convinced it's more readable than what's there already. Is there any advantage to using `.head<3>()` instead of `.head(3)`?
The docs just say the templated one is fixed size, and the untemplated one is dynamic - does that mean we save some overhead at runtime?

Reply:
`head<3>` should save us runtime overhead. @TrentHouliston what is your opinion on this matter?

Reply:
A number of advantages to head<3>() if it works.
Downside is it's not dynamic so you can't change the 3, but it's hardcoded here anyway.
On the plus side is the vector that comes out is a fixed size. That means if you're doing matrix multiplication etc it can check sizes and throw errors at compile time.
It'll also be slightly faster when combined with eigen's expression trees etc. Dynamic matrices need to allocate dynamic memory (new/malloc) while fixed size ones go on the stack. That's actually the biggest performance improvement here. Dynamic memory is slow...

Reply:
Roger that. I'll convert each of them to that type then.